gpytorch/lazy/diag_lazy_tensor.py

```diff
@@ -3,159 +3,131 @@
 
     def inv_quad_log_det(self, inv_quad_rhs=None, log_det=False, reduce_inv_quad=True):
-        if inv_quad_rhs is not None:
-            if (self.dim() == 2 and inv_quad_rhs.dim() == 1):
-                inv_quad_rhs = inv_quad_rhs.unsqueeze(-1)
-                if self.shape[-1] != inv_quad_rhs.numel():
-                    raise RuntimeError(
-                        "LazyTensor (size={}) cannot be multiplied with right-hand-side Tensor (size={}).".format(
-                            self.shape, inv_quad_rhs.shape
-                        )
-                    )
-            elif self.dim() != inv_quad_rhs.dim():
-                raise RuntimeError(
-                    "LazyTensor (size={}) and right-hand-side Tensor (size={}) should have the same number "
-                    "of dimensions.".format(self.shape, inv_quad_rhs.shape)
-                )
-            elif self.batch_shape != inv_quad_rhs.shape[:-2] or self.shape[-1] != inv_quad_rhs.shape[-2]:
-                raise RuntimeError(
-                    "LazyTensor (size={}) cannot be multiplied with right-hand-side Tensor (size={}).".format(
-                        self.shape, inv_quad_rhs.shape
-                    )
-                )
-
-        inv_quad_term = None
+
         if inv_quad_rhs is None:
-            inv_quad_term = torch.tensor([], dtype=self.dtype, device=self.device)
+            inv_quad_term = torch.empty(0, dtype=self.dtype, device=self.device)
         else:
-            inv_quad_term = inv_quad_rhs.div(self._diag.unsqueeze(-1)).mul(inv_quad_rhs).sum(-2)
+            if self._diag.shape[-1] != inv_quad_rhs.shape[-1]:
+                raise Exception
+            inv_quad_term = inv_quad_rhs.div(self._diag).mul(inv_quad_rhs).sum(-1)
             if reduce_inv_quad:
                 inv_quad_term = inv_quad_term.sum(-1)
 
-        log_det_term = None
-        if log_det:
-            log_det_term = self._diag.log().sum(-1)
+        if not log_det:
+            log_det_term = torch.empty(0, dtype=self.dtype, device=self.device)
         else:
-            log_det_term = torch.tensor([], dtype=self.dtype, device=self.device)
+            log_det_term = self._diag.log().sum(-1)
 
         return inv_quad_term, log_det_term
 
+    def log(self):
+        return DiagLazyTensor(self._diag.log())
+
     def root_decomposition(self):
-        if self.exact_root_decomposition:
-            return self.__class__(self._diag.sqrt()).evaluate()
-        else:
-            return super(DiagLazyTensor, self).root_decomposition()
+        return DiagLazyTensor(self._diag.sqrt())
 
     def root_inv_decomposition(self, initial_vectors=None, test_vectors=None):
-        if self.exact_root_decomposition:
-            return self.__class__(self._diag.sqrt().reciprocal()).evaluate()
-        else:
-            return super(DiagLazyTensor, self).root_decomposition()
+        return self.root_decomposition().inverse()
+
+    def sum_batch(self, sum_batch_size=None):
```

Comment:
This should be easy to define / shouldn't need to raise a `NotImplementedError`. Just sum `self._diag` along all batch dimensions and return a new `DiagLazyTensor`.

Reply:
yeah I just didn't get to this thus far and so `NotImplementedError` seemed to be the right error to raise :)

Reply:
So I assume the reason `SumBatchLazyTensor` exists at all is because gpytorch uses a single batch to represent a batch of blocks. Is there a particular reason this was done (apart form the lack of support for multi-dimensional batch shapes)? I feel this should be moved to a proper multi-shape representation (e.g. `b x k x n x n` instead of `bk x n x n`)

Reply:
It's how we handle additive structure efficiently in the setting where the kernels we are adding can be coalesced in to a single batch tensor or lazy tensor (see AdditveStructureKernel)

Reply:
Specifically, SumBatchLazyTensor's matmul is far more efficient than SumLazyTensor's

Reply:
Sure that makes sense. I guess what I was asking about is the internal representation as a multi-dim batch (vs. a single-dim batch with batch size `bk`). That shouldn't really affect perf much, but avoid some size arithmetic. Not important at this point though, just a side note

Reply:
Oh, yeah. I think moving forward that `b x k x n x n` makes the most sense.